pybossa/sched.py

```diff
@@ -590,20 +591,31 @@ def acquire_lock(task_id, user_id, limit, timeout, pipeline=None, execute=True):
     return False
 
 
-def release_reserve_task_lock_by_id(project_id, task_id, user_id, timeout, expiry=EXPIRE_RESERVE_TASK_LOCK_DELAY):
+def release_reserve_task_lock_by_id(project_id, task_id, user_id, timeout, expiry=EXPIRE_RESERVE_TASK_LOCK_DELAY, release_all_task=None):
     reserve_key = get_reserve_task_key(task_id)
     if not reserve_key:
         return
 
     redis_conn = sentinel.master
     lock_manager = LockManager(redis_conn, timeout)
-    resource_id = "reserve_task:project:{}:category:{}:user:{}:task:{}".format(
-        project_id, reserve_key, user_id, task_id)
-    lock_manager.release_reserve_task_lock(resource_id, expiry)
-    current_app.logger.info(
-        "Release reserve task lock. project %s, task %s, user %s, expiry %d",
-        project_id, task_id, user_id, expiry
-    )
+    if release_all_task:
+        pattern = "reserve_task:project:{}:category:{}:user:{}:task:*".format(
+            project_id, reserve_key, user_id)
+        resource_ids = lock_manager.scan_keys(pattern)
+        for k in resource_ids:
+            task_id_in_key = int(k.decode().split(":")[-1])
+            # If a task is locked by the user(in other tab), then the category lock should not be released
+            if task_id_in_key == task_id or not lock_manager.has_lock(get_task_users_key(task_id_in_key), user_id):
+                lock_manager.release_reserve_task_lock(k, expiry)
+                current_app.logger.info("Release reserve task locks: %s, task %d", k, task_id_in_key)
```

Comment:
Minor, perhaps add `project_id` and `user_id` to the log as well? 

Additionally, I guess we want to make this log different than the one at line 615 so that we can distinguish which code is executed, maybe add `release_all_task=True` into log as well or aggregate all released categories in a single log message?

Reply:
Sure, will do.